src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+//! Utility to gracefully shutdown a server.
+//!
+//! This module provides a [`GracefulShutdown`] type,
+//! which can be used to gracefully shutdown a server.
+//!
+//! See <https://github.com/hyperium/hyper-util/blob/master/examples/server_graceful.rs>
+//! for an example of how to use this.
+
+use pin_project_lite::pin_project;
+use slab::Slab;
+use std::{
+    fmt::{self, Debug},
+    future::Future,
+    pin::Pin,
+    sync::{atomic::AtomicUsize, Arc, Mutex},
+    task::{self, Poll, Waker},
+};
+
+/// A graceful shutdown watcher
+pub struct GracefulShutdown {
+    // state used to keep track of all futures that are being watched
+    state: Arc<GracefulState>,
+    // state that the watched futures to know when shutdown signal is received
+    future_state: Arc<GracefulState>,
```

Comment:
IIUC, the future_state here is to keep track of all the futures. What does the `state` do for GracefulShutdown here?

Reply:
IIUC = If I understand Correctly, didn't know that was a thing.

Reply:
Naming is hard though. I am essentially using the same concept in two directions.

- the "state" is to keep track of all watchers active. When the `GracefulShutdown` turns into a `GracefulWaiter` upon `self::shutdown()` it uses it to check when the counter is 0 and thus when shutdown is finished. While it is still a `GracefulShutdown` it clones this state and patches it to the `GracefulWatchers` (the ones watching the "graceful" connections) to help to do its bookkeeping using this counter
- the "future_state" is only ever once incremented (it is the counter of the `GracefulShutdown` itself, of which there is only ever one. And it is used by the "graceful" connections to know if a cancel is triggered. This is essentially "reusing" the `GracefulWaiter` logic that was already created for the other direction.

So no probably due to the fault of my part (not sure if bad documentation, naming or code) it seemed that you didn not understand it correctly. As it is the `state` which is there to keep track of all the futures (futures as in graceful connections).